python/bindings.cpp

```diff
@@ -375,13 +375,45 @@ PYBIND11_MODULE(_libsonata, m) {
     py::class_<Selection>(m,
                           "Selection",
                           "ID sequence in the form convenient for querying attributes")
-        .def(py::init<const Selection::Ranges&>(), "ranges"_a, "Selection from list of intervals")
-        .def(py::init([](py::array_t<uint64_t, py::array::c_style | py::array::forcecast> values) {
+        .def(py::init<const Selection::Ranges&>(),
+             py::arg("ranges"),
+             "Selection from list of intervals")
+        .def(py::init([](py::array_t<uint64_t, py::array::c_style> values) {
                  const auto raw = values.unchecked<1>();
                  return Selection::fromValues(raw.data(0), raw.data(raw.shape(0)));
              }),
+             py::arg("values").noconvert(true),
+             "Selection from list of IDs: passing np.array with dtype np.uint64 is faster")
+        .def(py::init([](py::array_t<int64_t, py::array::c_style | py::array::forcecast> values) {
+                 /* Both the Selection::Range and fromValues cases are handled here:
+                  * It doesn't seem possible to prevent the cast of [(-1, 1), (2, 3) ... ] style
+                  * Ranges; the Numpy auto-conversion from happening in this case.
+                  */
+                 py::buffer_info info = values.request();
+                 if (info.ndim == 2) {
+                     const auto raw = values.unchecked<2>();
+                     Selection::Ranges ranges;
+                     ranges.reserve(raw.shape(0));
+                     for (py::ssize_t i = 0; i < raw.shape(0); ++i) {
+                         if (raw(i, 0) < 0 || raw(i, 1) < 0) {
+                             throw SonataError("Negative value passed to Selection");
+                         }
+                         ranges.emplace_back(raw(i, 0), raw(i, 1));
+                     }
+                     return Selection(ranges);
+                 }
+                 // one dimensional case; ie fromValues
+                 const auto raw = values.unchecked<1>();
+                 for (size_t i = 0; i < raw.shape(0); ++i) {
+                     if (raw[i] < 0) {
+                         throw SonataError("Negative value passed to Selection");
+                     }
+                 }
+
+                 return Selection::fromValues(raw.data(0), raw.data(raw.shape(0)));
+             }),
              "values"_a,
-             "Selection from list of IDs")
+             "Selection from list of IDs: passing np.array with dtype np.uint64 is faster")
```

Comment:
Is this quantifiable?

Reply:
Wouldn't have written it otherwise:
```
In [1]: import numpy as np
In [2]: import libsonata
In [3]: uint64 = np.arange(1000000, dtype=np.uint64)
In [4]: int64 = np.arange(1000000, dtype=np.int64)
In [5]: a = range(1000000)
In [6]: %timeit libsonata.Selection(uint64)
2.07 ms ± 11.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
In [7]: %timeit libsonata.Selection(int64)
20.3 ms ± 68.7 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
In [8]: %timeit libsonata.Selection(a)
143 ms ± 525 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

pybind11's conversion machinenery kicks in when the types don't match, causing a conversion and thus the overhead.

Reply:
nice!